src/main/java/org/elasticsearch/gateway/GatewayMetaState.java

```diff
@@ -219,6 +217,22 @@ public void clusterChanged(ClusterChangedEvent event) {
 
             // check and write changes in indices
             for (IndexMetaData indexMetaData : newMetaData) {
+
+                boolean shardsAllocatedOnThisNodeInLastClusterState = true;
+                if (isDataOnlyNode(state)) {
```

Comment:
I don't think we need to check this everytime. we can just assign a boolean in the ctor if we need to write the state of not and if we need to write all states or not and centralize the decision. Then we can also remove the `isDataOnlyNode` method